capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+                self.read_bytes(write_buffer, read_buffer, 1);
+            }
+
+            State::WriteBlockBusy => {
+                if (read_buffer[0] & 0x1F) == 0x05 {
+                    // check if sd card is busy
+                    self.state.set(State::WaitWriteBlockBusy);
+                    self.read_bytes(write_buffer, read_buffer, 1);
+                } else {
+                    // error, send callback and quit
+                    self.txbuffer.replace(write_buffer);
+                    self.rxbuffer.replace(read_buffer);
+                    self.state.set(State::Idle);
+                    self.alarm_state.set(AlarmState::Idle);
+                    self.alarm_count.set(0);
+                    self.client
+                        .get()
+                        .map(move |client| { client.error(ErrorCode::WriteFailure as u32); });
+                }
+            }
+
+            State::WaitWriteBlockBusy => {
+                if read_buffer[0] != 0x00 {
+                    // replace buffers
+                    self.txbuffer.replace(write_buffer);
+                    self.rxbuffer.replace(read_buffer);
+
+                    // read finished, perform callback
+                    self.state.set(State::Idle);
+                    self.alarm_count.set(0);
+                    self.client_buffer.take().map(move |buffer| {
+                        self.client.get().map(move |client| { client.write_done(buffer); });
+                    });
+                } else {
+                    // replace buffers
+                    self.txbuffer.replace(write_buffer);
+                    self.rxbuffer.replace(read_buffer);
+
+                    // try again after 1 ms
+                    self.alarm_state.set(AlarmState::WaitForWriteBusy);
+                    let interval = (1 as u32) * <A::Frequency>::frequency() / 1000;
+                    let tics = self.alarm.now().wrapping_add(interval);
+                    self.alarm.set_alarm(tics);
+                }
+            }
+
+            State::Idle => {
+                // receiving an event from Idle means something was killed
+
+                // replace buffers
+                self.txbuffer.replace(write_buffer);
+                self.rxbuffer.replace(read_buffer);
+            }
+        }
+    }
+
+    // updates SD card state upon timer alarm fired
+    fn process_alarm_states(&self) {
+        // keep track of how many times the alarm has been called in a row
+        let repeats = self.alarm_count.get();
+        if repeats > 100 {
+            // error, send callback and quit
+            self.state.set(State::Idle);
+            self.alarm_state.set(AlarmState::Idle);
+            self.alarm_count.set(0);
+            self.client
+                .get()
+                .map(move |client| { client.error(ErrorCode::TimeoutFailure as u32); });
+        } else {
+            self.alarm_count.set(repeats + 1);
+        }
+
+        match self.alarm_state.get() {
```

Comment:
I'd accept suggestions. Mostly those checks were for debugging, but if we ever reach one of these timer callbacks without having called `replace` on the buffers the whole driver will fall apart.